repl/option.go

```diff
@@ -0,0 +1,116 @@
+package repl
+
+import (
+	"bufio"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/spy16/sabre"
+)
+
+// Option implementations can be provided to New() to configure the REPL
+// during initialization.
+type Option func(repl *REPL)
+
+// ReaderFactory should return an instance of reader when called. This might
+// be called repeatedly. See WithReaderFactory()
+type ReaderFactory func(r io.Reader) *sabre.Reader
```

Comment:
IMHO this should be an interface.  The logic to set up reader macros can easily become complex, especially when you're looking to re-use objects across Readers or when your language has primitives that deal with the network (as is my case with wetware).

For example, my `ReaderFactory` will undoubtedly be a struct that encapsulates a libp2p `host.Host` instance, which then needs to be bound to the `Anchor` macro.  It'll end up looking something like this:

```go
// Register the wetware Anchor macro.  This is directly copy-pasted
// from the existing ww source.
func Register(rd *sabre.Reader, root ww.Anchor) {
	s.SetMacro('/', anchorMacro(root), false)
}

type factory struct {
    log log.Logger
    host host.Host  // libp2p host
}

func (f factory) NewReader(r io.Reader) *sabre.Reader {
    root := anchor.New(f.host, anchor.WithLogger(f.log))

    rd := sabre.NewReader(r)
    Register(rd, root)
    
    return rd
}
```
Obviously this _could_ be implemented without the `readerFactory` struct, but this is a case where I anticipate wanting the encapsulation that structs provide.

In any case, I'd really like to be able to write:

```go
WithReaderFactory(factory{log: logger, host: h})
```

Reply:
Ah, fair enough. I actually looked at `NewReader` and made the `ReaderFactory` to simply match that. I guess we could have `ReaderFactory` as interface and may be provide a `ReaderFactoryFunc` also for convinience.

Reply:
Yup, that would be lovely.